backend/src/app/message.rs

```diff
@@ -329,6 +357,79 @@ pub async fn handle_app_message(mut app: App) {
                     status: lobby.status.clone(),
                 });
             }
+            AppMessage::ComputePlayerProgress {
+                lobby_id,
+                player_id,
+                progress,
+            } => {
+                let Some(lobby) = app.lobbies.get_mut(&lobby_id) else {
+                    error!("Lobby with ID {} was not found.", lobby_id);
+                    continue;
+                };
+                let Some(player) = lobby.players.get_mut(&player_id) else {
+                    error!(
+                        "Player with ID {} was not found in lobby {}.",
+                        player_id, lobby.name
+                    );
+                    continue;
+                };
+
+                // We only allow players to progress when the lobby is currently
+                // in progress.
+                let LobbyStatus::InProgress(_) = lobby.status else {
+                    warn!(
+                        "Player {} tried to progress in lobby {} that is not in progress.",
+                        player.name, lobby.name
+                    );
+                    continue;
+                };
+                let goal_file = match std::str::from_utf8(&lobby.challenge_files.goal_file) {
+                    Ok(goal_file) => goal_file,
+                    Err(e) => {
+                        error!("Error converting goal file bytes to string: {e}");
+                        continue;
+                    }
+                };
+                let player_file = match std::str::from_utf8(&progress) {
+                    Ok(player_file) => player_file,
+                    Err(e) => {
+                        error!("Error converting player file bytes to string: {e}");
+                        continue;
+                    }
+                };
+
+                // Compute the levenshtein distance between goal and player
+                // file.
+                let progress = normalized_levenshtein(goal_file, player_file);
+                player.progress = progress;
+
+                // If a player won we reduce the lobby lifetime and tell all
+                // players about it.
+                if progress.eq(&1.0) {
+                    lobby.status = LobbyStatus::InProgress(Utc::now() + REDUCED_LOBBY_PLAY_TIME);
+                    let app_tx = app.tx.clone();
+                    tokio::spawn(async move {
+                        tokio::time::sleep(REDUCED_LOBBY_PLAY_TIME).await;
+                        let _ = app_tx.send(AppMessage::Finish { lobby_id });
+                    });
+
+                    // Tell players that a player has won.
+                    let message = format!("Player {} finished first!", player.name);
+                    lobby.broadcast(BackendMessage::SendMessage(message));
```

Comment:
If another player finishes it will also say that this one finished first. Adjust the value accordingly.